File: 정종길/프로그래머스/카카오 2회독/2021 KAKAO BLIND RECRUITMENT/광고 삽입.py
def solution(play_time, adv_time, logs):
    ps, ads = toSec(play_time), toSec(adv_time) - 1
    prefix_sums = [0] * (ps + 1)

    for log in logs:
        s, e = log.split("-")
        prefix_sums[toSec(s)] += 1
        prefix_sums[toSec(e)] -= 1
    
    for i in range(ps):
        prefix_sums[i + 1] += prefix_sums[i]
    
    answer = 0
    i, j = 0, ads
    max_v = tot = sum(prefix_sums[i : j + 1])
    
    while j < ps:
        j += 1
        tot  += prefix_sums[j] - prefix_sums[i]
        i += 1
        if tot > max_v:
            max_v = tot
            answer = i
            
    hh, mm, ss = tohhmmss(answer)
    return f'{hh:02d}:{mm:02d}:{ss:02d}'

def toSec(txt):
    hh, mm, ss = txt.split(":")
    return int(hh) * 3600 + int(mm) * 60 + int(ss)

def tohhmmss(second):
    hh = second // 3600
    mm = (second % 3600) // 60
    ss = second % 60
    return hh, mm, ss

File: 정종길/프로그래머스/카카오 1회독/2021 KAKAO BLIND RECRUITMENT/광고 삽입.java
import java.util.*;

class Solution {
    public String solution(String play_time, String adv_time, String[] logs) {
        int playSec = toSeconds(play_time);
        int advSec = toSeconds(adv_time);
        int[] records = new int[playSec + 1];

        for(String log : logs){
            String[] logLi = log.split("-");
            int startSec = toSeconds(logLi[0]);
            int endSec = toSeconds(logLi[1]);
            records[startSec] += 1;
            records[endSec] -= 1;
        }

        for(int i = 1; i < records.length; i ++){
            records[i] += records[i - 1];
        }


        int[] subLi = Arrays.copyOfRange(records, 0, advSec + 1);
        long maxRunningTime = Arrays.stream(subLi)
                .sum();

        int start_ad = 0;
        int end_ad = advSec;
        long runningTime = maxRunningTime;
        int answer = 0;

        for(int k = advSec + 1; k < playSec + 1; k++){
            runningTime -= records[start_ad];
            runningTime += records[k];
            start_ad += 1;
            if (maxRunningTime < runningTime){
                answer = start_ad;
                maxRunningTime = runningTime;
            }
        }

        String hour = String.format("%02d", answer / 3600);
        String min = String.format("%02d", (answer % 3600) / 60);
        String sec = String.format("%02d", (answer % 3600) % 60);

        return hour + ":" + min + ":" + sec;
    }

    public int toSeconds(String text){
        int sec = 0;
        String[] times = text.split(":");

        sec += Integer.parseInt(times[0]) * 3600 + Integer.parseInt(times[1]) * 60 + Integer.parseInt(times[2]);
        return sec;
    }
}